### PR TITLE
Refactor and match and search mechanism

### PR DESCRIPTION
**Pull request by theosev**

### CODE REVIEW

File: services/search/mainsearch.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+				return
+			}
+
+			fmt.Println("There was a match")
+
+			insertEventsQuery := `INSERT INTO events ("aggregate_uuid", "aggregate_type", "payload", "payload_type", "version") VALUES ($1, $2, $3, $4, $5)`
+
+			for _, user := range users {
+				// Update user event
+				_, err = tx.ExecContext(ctx, insertEventsQuery, user.userID, "user", "{ \"status\": \"busy\" }", "user_status_updated", 3)
```
Comment (Dev-Force) [inline services/search/mainsearch.go:86]: Care for hardcoded values

File: services/search/mainsearch.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+// User describes user properties
+type User struct {
+	userID guuid.UUID
+	status string
+}
+
+// Run executes the logic of MainSearch
+func (ms *MainSearch) Run() {
+	for {
+		func() {
```
Comment (Dev-Force) [inline services/search/mainsearch.go:31]: Is there any reason to wrap this code inside a function?
I suppose the reason is the defer function.

File: services/search/mainsearch.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+			}
+			m := match.NewMainMatch(eOne, eTwo)
+			go m.Start()
+		}()
+		time.Sleep(2 * time.Second)
+	}
+}
+
+// AddUser adds a user in UserChannels
+func (ms *MainSearch) AddUser(userID guuid.UUID, uic UserMatchCom) error {
```
Comment (Dev-Force) [inline services/search/mainsearch.go:116]: We need synchronized access for the reads as well.
Like a GetUser function.
Ref: https://stackoverflow.com/a/11063566

File: services/search/mainsearch.go (hunk 2 above)
Comment (theosev) [reply services/search/mainsearch.go:31]: Exactly. Ofc we can export the code to private function

**Pull request merged**